scripts/utils.py

```diff
@@ -0,0 +1,134 @@
+#!/usr/bin/env python3
+
+"""
+Authors: Tobias Kaiser, Jakob Beckmann
+
+Copyright 2018 ChainSecurity AG
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+"""
+
+import sys
+import os
+import re
+import logging
+
+from solc.exceptions import SolcError
+import solc.install
+
+from termcolor import colored
+
+class NoSolidityProject(BaseException):
+    def __init__(self, dir):
+        self.dir = dir
+
+
+class CompilerVersionNotSupported(BaseException):
+    def __init__(self, version, too_old=True):
+        self.version = version
+        self.too_old = too_old
+
+
+class SolidityCompilationException(SolcError):
+    def __init__(self, solc_exception, files):
+        super().__init__(solc_exception.command,
+                         solc_exception.return_code,
+                         solc_exception.stdin_data,
+                         solc_exception.stderr_data,
+                         solc_exception.stdout_data,
+                         solc_exception.message)
+        self.files = files
+
+
+def get_supported_solc_versions():
+    """Gets the supported solc versions that can be installed from `pysolc`."""
+    versions = [getattr(solc.install, item)
+                for item in dir(solc.install) if item.startswith('V')]
+    versions = [v[1:] for v in versions]
+    return filter(lambda x: version_to_tuple(x) >= version_to_tuple('0.4.11'), versions)
+
+
+def version_to_tuple(v):
+    """Converts a version string into a tuple."""
+    return tuple(map(int, v.split('.')))
+
+
+def find_node_modules_dir(directory):
+    """Returns the path to the node module directory contained in the `directory` directory."""
+    for x in os.walk(directory):
+        if os.path.isdir(os.path.join(x[0], 'node_modules')):
+            return os.path.join(x[0], 'node_modules')
+    return None
+
+
+def parse_sol_version(source):
+    """Parses the solidity version from a contract using the pragma."""
+    with open(source, encoding='utf-8') as f:
+        for l in f.readlines():
```

Comment:
`lines = f.readlines()` to close the file earlier (and remove one indentation level).